base_report_to_printer/models/printing_server.py

```diff
@@ -0,0 +1,233 @@
+            existing_printers = dict([
+                (printer.system_name, printer)
+                for printer in server.printer_ids
+            ])
+            updated_printers = []
+            for name, printer_info in printers.iteritems():
+                printer_values = {
+                    'name': printer_info['printer-info'],
+                    'system_name': name,
+                    'server_id': server.id,
+                    'uri': printer_info.get('device-uri', ''),
+                    'status': mapping.get(printer_info.get('printer-state'),
+                                          'unknown'),
+                    'status_message': printer_info.get(
+                        'printer-state-message', ''),
+                    'model': printer_info.get('printer-make-and-model', False),
+                    'location': printer_info.get('printer-location', False),
+                }
+                updated_printers.append(name)
+                if name not in existing_printers.keys():
+                    self.env['printing.printer'].sudo().create(printer_values)
+                else:
+                    existing_printers[name].sudo().write(printer_values)
+
+            # Set printers not found as unavailable
+            server.printer_ids.filtered(
+                lambda record: record.system_name not in updated_printers)\
+                .write({'status': 'unavailable'})
+
+        return res
+
+    @api.multi
+    def action_update_jobs(self):
+        if not self:
+            self = self.search([])
+        return self.update_jobs()
+
+    @api.multi
+    def update_jobs(self, which='all', first_job_id=-1):
+        job_obj = self.env['printing.job'].sudo()
+        printer_obj = self.env['printing.printer']
+
+        mapping = {
+            3: 'pending',
+            4: 'pending held',
+            5: 'processing',
+            6: 'processing stopped',
+            7: 'canceled',
+            8: 'aborted',
+            9: 'completed',
+        }
+
+        # Update printers list, to ensure that jobs printers will be in Odoo
+        self.update_printers()
+
+        for server in self:
+            connection = server._openConnection()
+            if not connection:
+                continue
+
+            # Retrieve asked job data
+            jobs_data = connection.getJobs(
+                which_jobs=which, first_job_id=first_job_id,
+                requested_attributes=[
+                    'job-name',
+                    'job-id',
+                    'printer-uri',
+                    'job-media-progress',
+                    'time-at-creation',
+                    'job-state',
+                    'job-state-reasons',
+                    'time-at-processing',
+                    'time-at-completed',
+                ])
+
+            # Retrieve known uncompleted jobs data to update them
+            if which == 'not-completed':
+                min_jobs = job_obj.search([
```

Comment:
Yes, `min` means that we search for the oldest : `order='jobid'`.
This is used in CUPS' `getJob` method arguments as a filter (`first_job_id=min_job.jobid`), there is no need to retrieve the full list, but only jobs not known as completed in Odoo's database.